### PR TITLE
Backwards compatible fixes to bpf() syscall emulation

### PR DESCRIPTION
**Pull request by lmb**
bpf(): do not return errors via errno

    The Linux ABI returns all syscall errors via the function return value, not
    via errno.

    Fixes https://github.com/microsoft/ebpf-for-windows/issues/3749

Allow detecting if bpf() command is not implemented

    Use SetLastError to indicate to callers that a bpf() command is not 
    implemented. This avoids polluting the bpf() return value with platform
    specific error returns while still allowing detection of this important
    case.

Add forwards and backwards compatibility to bpf() emulation

    On Linux, bpf() accepts a bpf_attr which is larger than what the syscall
    expects, as long as the unknown fields are all 0. It also accepts a bpf_attr
    which is smaller than what it expects, by assuming that all missing fields
    are zero.

    This allows forwards and backwards compatilibity between old and new 
    versions of both the Linux kernel and user space tooling.

    Implement a similar scheme for the bpf() emulation.

Return EPERM from bpf() if user is not privileged

    On Linux, bpf() returns EPERM if the user doesn't have CAP_BPF. Return the
    same error when the user isn't able to open the device handle.

### CODE REVIEW

File: libs/api/bpf_syscall.cpp [hunk 1]
```diff
@@ -3,96 +3,157 @@
 #include "bpf.h"
 #include "libbpf.h"
 
+#include <windows.h>
+#include <WinError.h>
```
Comment (dthaler) [inline libs/api/bpf_syscall.cpp:7]: list includes in alphabetical order unless you get build errors.  If order does matter, add a // comment saying so.

File: libs/api/bpf_syscall.cpp [hunk 2]
```diff
@@ -3,96 +3,157 @@
...
+    int retval;
+
+    if (size > sizeof(tmp)) {
+        // Forward compatibility: allow a larger input as long as the
+        // unknown fields are all zero.
+        if (!tail_is_zero(attr, sizeof(tmp), size)) {
+            return -EINVAL;
+        }
+    } else {
+        // Backwards compatibility: allow a smaller input by implictly zeroing all
```
Comment (dthaler) [inline libs/api/bpf_syscall.cpp:52]:
```suggestion
        // Backwards compatibility: allow a smaller input by implicitly zeroing all
```
typo

**Review by dthaler — COMMENTED** [review]
Looks fine other than two style nits

File: libs/api/bpf_syscall.cpp (hunk 1 above)
Comment (lmb) [reply libs/api/bpf_syscall.cpp:7]: This sorting is enforced by clang-format as far as I can tell, which gives the windows headers higher priority.

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**